### PR TITLE
CSPL-763 - App framework related changes

### PR DESCRIPTION
**Pull request by kumarajeet**
**Overview** 
This PR is for a story as part of AppFramework Operator-Setup. This story provides configuration section to pass the App Framework related parameters. 

It is applicable to following CRs:
Standalone, ClusterMaster, SearchHeadCluster and LicenseMaster

By default, this feature is **disabled** unless we explicitly enable it by adding following under "applications" section

`featureEnabled: TRUE`

Currently unit tests are present for the configuration values for :
```
   appFrameworkRef:
     type: s3
     s3Endpoint: "http://10.244.0.6:9000"
     s3Bucket: "team1cluster"
     s3SecretRef: mysecret
     s3PollInterval: 60 
```

### CODE REVIEW

File: pkg/splunk/enterprise/configuration_test.go [hunk 1]
```diff
@@ -596,6 +596,85 @@ func TestValidateSplunkSmartstoreSpec(t *testing.T) {
 	}
 }
 
+
+
+func TestValidateApplicationFrameworkSpec(t *testing.T) {
+	var err error
+	// Valid app framework config with proper inputs
```
Comment (smohan-splunk) [inline pkg/splunk/enterprise/configuration_test.go:603]: Can we add a Positive test case too?

Comment (kumarajeet) [reply pkg/splunk/enterprise/configuration_test.go:603]: Sure, let me check.

File: pkg/apis/enterprise/v1beta1/common_types.go [hunk 2]
```diff
@@ -177,3 +177,26 @@ type IndexAndCacheManagerCommonSpec struct {
 	// Time period relative to the bucket's age, during which the bloom filter file is protected from cache eviction
 	HotlistBloomFilterRecencyHours uint `json:"hotlistBloomFilterRecencyHours,omitempty"`
 }
+
+// ApplicationFrameworkSpec defines the application package remote store repository
+type ApplicationFrameworkSpec struct {
+
+	// Is Application Framework Feature Enabled. 
+	// Default value is False
+	FeatureEnabled string `json:"featureEnabled"`
```
Comment (akondur) [inline pkg/apis/enterprise/v1beta1/common_types.go:186]: Any reason why a boolean wouldn't work here?

**Review by akondur — COMMENTED** [review]
Looks good. Couple of minor comments.

**Review by kashok-splunk — APPROVED** [review]

File: pkg/apis/enterprise/v1beta1/clustermaster_types.go [hunk 3]
```diff
@@ -38,6 +38,11 @@ type ClusterMasterSpec struct {
 	// Also used on cluster master for init container to setup the soft links from ../master-apps/splunk-operator/local/ to
 	// /mnt/splunk-operator/local/
 	SparkImage string `json:"sparkImage"`
+
+	// App Framework configuration. Refers to the config block for App Framework.
+	// Through this config, apps can be installed in an Indexer Cluster.
+	// The implementatiom is still TBD.
+	ApplicationFrameworkRef ApplicationFrameworkSpec `json:"applications"`
```
[the remark this answers is not in the document]
Comment (kumarajeet) [reply pkg/apis/enterprise/v1beta1/clustermaster_types.go]: this is changed now

File: pkg/apis/enterprise/v1beta1/common_types.go (hunk 2 above)
Comment (kumarajeet) [reply pkg/apis/enterprise/v1beta1/common_types.go]: changed to bool

**Review by akondur — APPROVED** [review]

File: pkg/splunk/enterprise/configuration.go [hunk 4]
```diff
@@ -756,6 +757,51 @@ func AreRemoteVolumeKeysChanged(client splcommon.ControllerClient, cr splcommon.
 	return false
 }
 
+// ValidateAppFrameworkSpec checks and validates the Apps Frame Work config
+func ValidateAppFrameworkSpec(appFramework *enterprisev1.AppFrameworkSpec) error {
+
+	//log all the inputs for AppFrameworkSpec.
+	//TBD: Future - use logging toggle switch to reduce any excessive logging.
+
+	log.Info("App Framework Inputs",
```
Comment (akondur) [inline pkg/splunk/enterprise/configuration.go:766]: Nit: Just thought about it. Should we move this log below the if check:

```
if !appFramework.FeatureEnabled {
    return nil
}
```
Reason being we can avoid the log if app framework is not enabled?

Also the format used is to create a scopedLog locally & use that.

```
scopedLog := log.WithName("ValidateAppFrameworkSpec").WithValues(
		"FeatureEnabled", appFramework.FeatureEnabled,
		"Type", appFramework.Type,
		"S3Endpoint", appFramework.S3Endpoint,
		"S3Bucket", appFramework.S3Bucket,
		"S3SecretRef", appFramework.S3SecretRef,
		"S3PollInterval", appFramework.S3PollInterval)
scopedLog.Info("Validating app framework spec")
```

Comment (kumarajeet) [reply pkg/splunk/enterprise/configuration.go:766]: Reviewing it.

Comment (kumarajeet) [reply pkg/splunk/enterprise/configuration.go]: Ok this is taken care of.

File: deploy/crds/enterprise.splunk.com_clustermasters_crd.yaml [hunk 5]
```diff
@@ -614,6 +614,31 @@ spec:
...
+                    is False
+                  type: boolean
+                s3Bucket:
+                  description: App Package Remote Store Bucket
+                  type: string
+                s3Endpoint:
+                  description: App Package Remote Store Endpoint
+                  type: string
+                s3PollInterval:
+                  description: App Package Remote Store Polling interval
```
Comment (vebken-splunk) [inline deploy/crds/enterprise.splunk.com_clustermasters_crd.yaml:633]: We should probably add the units and default value here so users know what they are configuring.

Also I noticed that it looks like the other code is written with seconds in mind, from our earlier discussion did we want to go with minutes instead?

File: pkg/splunk/enterprise/configuration_test.go (hunk 1 above)
Comment (kumarajeet) [reply pkg/splunk/enterprise/configuration_test.go:601]: Positive check was also added.

**Pull request closed without merging**